### PR TITLE
core/evm: Set Indestructible Opcode (AA Work)

### PR DESCRIPTION
**Pull request by technicallyty**
progress on set indestructible opcode

### CODE REVIEW

File: core/blockchain_test.go [hunk 1]
```diff
@@ -2436,6 +2436,135 @@ func TestSideImportPrunedBlocks(t *testing.T) {
 	}
 }
```
Comment (lightclient) [inline core/blockchain_test.go:2438]: Would be nice to comment this and briefly describe the test.

File: core/blockchain_test.go [hunk 2]
```diff
@@ -2436,6 +2436,135 @@ func TestSideImportPrunedBlocks(t *testing.T) {
...
+	if err != nil {
+		t.Fatalf("failed to create tester chain: %v", err)
+	}
+	if n, err := chain.InsertChain(blocks); err != nil {
+		t.Fatalf("block %d: failed to insert into chain: %v", n, err)
+	}
+	st, _ := chain.State()
+
+	if !st.Exist(common.HexToAddress("0x000000000000000000000000000000000000bbbb")) {
+		t.Fatalf("Contract self destruct even with set_indestructible set.")
```
Comment (lightclient) [inline core/blockchain_test.go:2564]:
```suggestion
		t.Fatalf("Contract self-destructed even when set as indestructible.")
```

File: core/vm/instructions.go [hunk 3]
```diff
@@ -815,11 +815,14 @@ func opStop(pc *uint64, interpreter *EVMInterpreter, callContext *callCtx) ([]by
 }
 
 func opSuicide(pc *uint64, interpreter *EVMInterpreter, callContext *callCtx) ([]byte, error) {
-	beneficiary := callContext.stack.pop()
-	balance := interpreter.evm.StateDB.GetBalance(callContext.contract.Address())
-	interpreter.evm.StateDB.AddBalance(common.Address(beneficiary.Bytes20()), balance)
-	interpreter.evm.StateDB.Suicide(callContext.contract.Address())
-	return nil, nil
+	if !interpreter.indestructible {
```
Comment (lightclient) [inline core/vm/instructions.go:818]: I think it is a bit cleaner to have something like:

```go
if interpreter.indestructible {
  return nil, ErrContractIndestructible
}

...

return nil, nil
```

File: core/vm/instructions.go [hunk 4]
```diff
@@ -855,7 +858,7 @@ func opPush1(pc *uint64, interpreter *EVMInterpreter, callContext *callCtx) ([]b
 		codeLen = uint64(len(callContext.contract.Code))
 		integer = new(uint256.Int)
 	)
-	*pc += 1
+	*pc++
```
Comment (lightclient) [inline core/vm/instructions.go:861]: Probably shouldn't change this line as it is unrelated to the PR.

File: core/vm/interpreter.go [hunk 5]
```diff
@@ -47,7 +47,7 @@ type Config struct {
 type Interpreter interface {
 	// Run loops and evaluates the contract's code with the given input data and returns
 	// the return byte-slice and an error if one occurred.
-	Run(contract *Contract, input []byte, static bool) ([]byte, error)
+	Run(contract *Contract, input []byte, static bool, indestructible bool) ([]byte, error)
```
Comment (lightclient) [inline core/vm/interpreter.go:50]: Some go syntactic sugar might be nice:

```suggestion
	Run(contract *Contract, input []byte, static, indestructible bool) ([]byte, error)
```

File: core/vm/interpreter.go [hunk 6]
```diff
@@ -241,6 +249,11 @@ func (in *EVMInterpreter) Run(contract *Contract, input []byte, readOnly bool) (
 				return nil, ErrWriteProtection
 			}
 		}
+		if in.indestructible {
```
Comment (lightclient) [inline core/vm/interpreter.go:252]: This is very similar functionality to the check that `opSelfdestruct` does. Is it possible to rely only on one or the other? Is this statement catching a particular code path? If possible - I think it is best to rely on the implementation of this in `opSelfdestruct` so each opcode doesn't need to have a `operation.selfdestructs` flag.

File: core/vm/opcodes.go [hunk 7]
```diff
@@ -403,150 +405,151 @@ func (op OpCode) String() string {
...
+	"SWAP13":            SWAP13,
+	"SWAP14":            SWAP14,
+	"SWAP15":            SWAP15,
+	"SWAP16":            SWAP16,
+	"LOG0":              LOG0,
+	"LOG1":              LOG1,
+	"LOG2":              LOG2,
+	"LOG3":              LOG3,
+	"LOG4":              LOG4,
+	"SETINDESTRUCTIBLE": SETINDESTRUCTIBLE,
```
Comment (lightclient) [inline core/vm/opcodes.go:545]: It's really unfortunate that adding this line creates such a big diff.

**Review by lightclient — COMMENTED** [review]
Awesome work :) I left a few comments.

File: core/vm/interpreter.go (hunk 6 above)
Comment (astarinmymind) [reply core/vm/interpreter.go:252]: You're right. The operation.selfdestructs flag is unnecessary.

File: core/vm/instructions.go (hunk 4 above)
Comment (technicallyty) [reply core/vm/instructions.go:861]: yeah, i have a bad habit of cleaning all the things golinter complains about, I'll have to try to resist the urge to appease it lol
